Create `PauliProductMeasurement` instruction

ShellyGarion opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Adds rust and python support for `PauliProductMeasurement` instructions. This is a fundamental operation in fault-tolerant quantum computing [1].

Close #15104 
Co-authored by @alexanderivrii 

References: 

* Daniel Litinski. "A Game of Surface Codes: Large-Scale Quantum Computing with Lattice Surgery",  <https://arxiv.org/abs/1808.02892>

### Details and comments

The implementation is similar to the existing handling of unitary gates.

**In rust:**

* There is a new variant `PauliProductMeasurement` for `PackedOperationType`. This increases the total number of types to `7`.
* There is a new struct `PauliProductMeasurement` defined in `operations.rs`. Currently it stores 
```rust
/// This class represents a PauliProductMeasurement instruction.
#[derive(Clone, Debug)]
#[repr(align(8))]
pub struct PauliProductMeasurement {
    /// The z-component of the pauli.
    pub z: Vec<bool>,
    /// The x-component of the pauli.
    pub x: Vec<bool>,
    /// For a PauliProductMeasurement instruction, the phase of the Pauli can be either 0 or 2,
    /// where the value of 2 corresponds to a sign of `-1`.
    pub neg: bool,
}
```
but we are considering bitpacking the data structure.

* There is a new variant `PauliProductMeasurement ` of `OperationRef`.

* The `synthesis` crate contains a function `synthesize_ppm`, which is used in `HighLevelSynthesis` transpiler pass to synthesize `PauliProductMeasurement` instructions (this is very similar to handling unitary gates) and is also exposed to Python (allowing to extract `definition` on the python side).

**In python:**

* There is a "mirror" class `PauliProductMeasurement`, which can be instantiated from a `Pauli` object:
```python

ppm = PauliProductMeasurement(Pauli("-XYIZ"))

qc = QuantumCircuit(6, 2)
qc.append(ppm, [0, 1, 2, 3], [0])
```
* Internally, the pauli data `(pauli.z, pauli.x, pauli.phase)` is stored in the `params` field, allowing to easily pass it down to rust.

* QPY serialization works
* Visualization is done using the `label` attribute. Currently this just draws a box over relevant quantum and classical bits. 
* Default and custom labels are preserved correctly

Note that we support pauli product measurement instructions based on Paulis with some "I"-terms (except for all-"I"), since we believe that this might simplify user workflows. However, for efficiency it would be better for the users to avoid creating "I"s, and we will try to document this clearly both in the docstring and the release note

### ToDo (in this PR)

- [ ] Write release note once the implementation is a bit more stable.
- [x] Port the class to rust, the quantum circuit generation should be based on [Pauli Evolution](https://github.com/Qiskit/qiskit/blob/main/crates/circuit_library/src/pauli_evolution.rs).
- [x] Add Tests to check that transpiler passes that handle measurements do not fail.
- [x] Minor fix to visualization (to print “PauliProductMeasurement” and not “Pauliproductmeasurement”).
- [x] Handle  QPY serialization.
- [ ] Improve the PauliProductMeasurement data structure (at least in Rust).
- [x] Double-check that the endianness of Pauli is consistent with the one in `SparseObservable`.

### ToDo (in follow-ups)

- [ ] Let visualization draw the double-wire down to the classical register, like the standard measurements do.
- [ ] Update the `LitinskiTransformation` pass to produce PPMs in addition to PPRs (pauli product rotations).
- [ ] Rethink and improve transpiler passes involving measurement instructions
- [ ] Qasm?

ShellyGarion committed:
add PauliProductMeasurement class.
Co-authored-by: Alexander Ivrii <alexi@il.ibm.com>

ShellyGarion committed:
update init file

ShellyGarion committed:
add tests for pauli product measurement

ShellyGarion committed:
fix definition, handle I in the lsb qubits, fix test

alexanderivrii committed:
Implementing PauliProductMeasurement also in Rust

* Conversions between rust and python classes

* Synthesis of PauliProductMeasurement instructions

* Exposing synthesis function to python

* Usage of synthesis function in HLS

* QPY support

* Some additional python tests

alexanderivrii committed:
cleanup and more tests

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,138 @@
+        """
+        Args:
+            pauli: A tensor product of Pauli operators defining the measurement,
+                for example ``Pauli("XY")`` or ``Pauli("-XYIZ)``.
```
```suggestion
                for example ``Pauli("XY")`` or ``Pauli("-XYIZ")``.
```

alexanderivrii replied:
Done in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,138 @@
+        .. note::
+
+            While Paulis involving "I"-terms are fully supported, it is recommended to remove
+            "I"-terms from the Pauli when creating a ``PauliProductMeasurement`` instruction,
```
```suggestion
            While Paulis involving ``"I"``-terms are fully supported, it is recommended to remove
            ``"I"``-terms from the Pauli when creating a ``PauliProductMeasurement`` instruction,
```

alexanderivrii replied:
Done in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,138 @@
+
+        """
+
+        if not isinstance(pauli, Pauli):
```
We generally don't have these type checks in the circuit library -- the type hint is clear here so it's on the user to provide the right type.

alexanderivrii replied:
I think it could be nice to have this specific check. I can see users trying to create `PauliProductMeasurement` instructions from strings (labels) or some other data, and having an explicit error message could help.

Cryoris replied:
I'm fine keeping it as long as we don't type check when we construct PPMs from the Litinski transformation, otherwise we'll spend a lot of time checking data we know is valid. If we do have to call this constructor, then we could add a `skip_checks` argument or something.

ShellyGarion replied:
I don't think that there are furhter type checks in Litinski transformation.

alexanderivrii replied:
I think the Litinski transform in the follow up PR should use the `_from_pauli_data` constructor which avoids these checks (if not, I will make sure to update it).

Cryoris commented on test/python/circuit/library/test_ppm.py:
```diff
@@ -0,0 +1,189 @@
+        with self.assertRaises(CircuitError):
+            _ = PauliProductMeasurement(Pauli(p))
+
+    @data("", "II", "-III")
```
Arguably `""` is not the identity string 😉 A separate error here seems better.

alexanderivrii replied:
Addressed in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on test/python/circuit/library/test_ppm.py:
```diff
@@ -0,0 +1,189 @@
+        qc3.append(PauliProductMeasurement(Pauli("XZ")), [4, 1], [0])
+
+        qc4 = QuantumCircuit(5, 2)
+        qc4.append(PauliProductMeasurement(Pauli("ZX")), [4, 1], [1])
```
It would be good to add a case here that checks that reverting the Pauli labels and qubit indices evaluates to `True`.

alexanderivrii replied:
Hmm, this would indeed be nice, however this is not supported right now (pretty much the same way as a `CZ`/`RZZ` instruction over qubits `[0, 1]` is not equal to a `CZ`/`RZZ` instruction over `[1, 0]`). 

I don't know if we can always canonicalize `PauliProductMeasurement` (and other similarly symmetric) instructions when adding them to a circuit -- can it happen that `ZX` over `[0, 1]` is supported by `XZ` over `[1, 0]` is not?

Cryoris commented on crates/synthesis/src/pauli_product_measurement.rs:
```diff
@@ -0,0 +1,154 @@
+    let num_qubits = ppm.num_qubits();
+    let num_clbits = ppm.num_clbits();
+
+    let mut circuit = CircuitData::with_capacity(num_qubits, num_clbits, 0, Param::Float(0.0))
```
Since you're returning a `PyResult`, we can just handle all these errors, there's no point in unwrapping them here.

alexanderivrii replied:
Good point, addressed in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on crates/synthesis/src/pauli_product_measurement.rs:
```diff
@@ -0,0 +1,154 @@
+}
+
+#[pyfunction]
+fn synth_pauli_product_measurement(py: Python, py_op: Py<PyAny>) -> PyResult<CircuitData> {
```
We could just take a `Bound<PyAny>` here instead of the two arguments, that'll simplify the code a little and it's what we mostly do now I think

alexanderivrii replied:
Good point, addressed in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on crates/synthesis/src/pauli_product_measurement.rs:
```diff
@@ -0,0 +1,154 @@
+    // The code is somewhat similar to pauli_evolution synthesis, however combining the two branches
+    // is tricky.
+    let num_qubits = ppm.num_qubits();
+    let num_clbits = ppm.num_clbits();
```
This is always 1, we could avoid querying this, right?

alexanderivrii replied:
Good point, addressed in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on crates/circuit/src/operations.rs:
```diff
@@ -2890,3 +2903,82 @@ impl UnitaryGate {
+        // Similarly to UnitaryGate, we do not provide the actual decomposition here.
+        // Instead, the HighLevelSynthesis transpiler pass is modified to call the
+        // relevant synthesis function when requiring the definition for a
+        // PauliProudctMeasurement.
```
```suggestion
        // PauliProductMeasurement.
```

alexanderivrii replied:
Fixed in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on crates/circuit/src/operations.rs:
```diff
@@ -275,6 +277,7 @@ pub enum OperationRef<'a> {
     Instruction(&'a PyInstruction),
     Operation(&'a PyOperation),
     Unitary(&'a UnitaryGate),
+    PPM(&'a PauliProductMeasurement),
```
Could we spell `PPM` out? No other gates are abbreviated and it'll be more consistent with `PauliEvolutionGate` or `PauliRotationGate`.

alexanderivrii replied:
I have renamed `PPM` to `PauliProductMeasurement` in [1335e71](https://github.com/Qiskit/qiskit/pull/15126/commits/1335e7129b351fc9d70b11e3e6b57cfea3b1685f).

Cryoris commented on crates/circuit/src/operations.rs:
```diff
@@ -2890,3 +2903,82 @@ impl UnitaryGate {
+#[repr(align(8))]
+pub struct PauliProductMeasurement {
+    /// The z-component of the pauli.
+    pub z: Vec<bool>,
```
Are there alternatives to storing boolean vectors, i.e. 1 byte per term if we'd only need 1 bit? E.g. the `bitvec` package seems to have a convenient interface. I assume @jakelishman has some opinions here 🙂

mtreinish replied:
What about fixedbitset? https://docs.rs/fixedbitset/latest/fixedbitset/index.html We already use this in qiskit (and petgraph and rustworkx use it).

Cryoris commented on crates/circuit/src/operations.rs:
```diff
@@ -2890,3 +2903,82 @@ impl UnitaryGate {
+
+/// This class represents a PauliProductMeasurement instruction.
+#[derive(Clone, Debug)]
+#[repr(align(8))]
```
In the current format I don't think we need this, since bool and u8 are both a byte long. But if we change to `bitvec` or some bitpacking we might need to keep it?

Cryoris commented on crates/circuit/src/operations.rs:
```diff
@@ -2890,3 +2903,82 @@ impl UnitaryGate {
+/// This class represents a PauliProductMeasurement instruction.
+#[derive(Clone, Debug)]
+#[repr(align(8))]
+pub struct PauliProductMeasurement {
```
We'll need to construct the PPMs in the Rust Litinski transformation code, which currently outputs the Pauli in a sparse string format (sign, Paulistring, indices). If we use ZX bits as internal representation we'll also need a converter from Pauli string to ZX in Rust.

alexanderivrii replied:
I will try other internal formats and see what is the best representation when extending the `LitinskiTransformation` pass.

Cryoris commented on test/python/circuit/library/test_ppm.py:
```diff
@@ -0,0 +1,189 @@
+        with self.assertRaises(CircuitError):
+            _ = PauliProductMeasurement(Pauli("XYZ")).inverse()
+
+    def test_qpy(self):
```
+1 for thinking of qpy 👍🏻 I guess for qasm export it'll just define the instruction?

alexanderivrii replied:
Unfortunately qasm export of instructions with classical qubits is not supported. @jakelishman suggested not to worry about this for now.

Cryoris reviewed (commented):
The overall structure LGTM, my main comment is about the internal Pauli representation. Do we want to generally adopt a dense ZX formulation for Cliffords and PBC? If yes, we (1) need to ensure we can build the PPMs from the string output of the Litinski transformation, and (2) it would be nice to investigate bitpacking (e.g. `bitvec` package?).

alexanderivrii committed:
Addressing review comments:

renaming PPM -> PauliProductMeasurement

avoiding redundant unwraps

returning a different error when creating a PauliProductMeasurement from an empty Pauli

using Bound<PyAny>

adding tests for running various transpiler passes on ppm circuits

avoid querying num_clbits

typos

docstring fixes

alexanderivrii commented:
I have addressed most of comments by @Cryoris and have removed the draft/wip status to invite other people to look at this PR. The main remaining comment is deciding on the best rust representation of the `PauliProductMeasurement` struct (but note that we can also change this in a follow-up). Explicitly, I am now planning to work on extending the `LitinskiTransformation` pass to handle `PauliProductMeasurements`, hopefully this would offer additional insights on how we want this data structure to look like.

alexanderivrii committed:
Representing the phase of the PauliProductMeasurement instruction in Rust as a boolean

alexanderivrii committed:
Fixing a problem when converting a Rust PPM instruction to Python, and adding a test

alexanderivrii committed:
Adding release note

alexanderivrii committed:
Merge branch 'main' into ppm

alexanderivrii committed:
changes for pyo3 update

alexanderivrii committed:
tweak to release notes

mtreinish commented on crates/circuit/src/circuit_instruction.rs:
```diff
@@ -618,6 +622,51 @@ impl<'a, 'py> FromPyObject<'a, 'py> for OperationFromPython {
                     });
                 };
             }
+        } else if ob_name == "PauliProductMeasurement" {
```
The gate name is camel case? I would have expected `"pauli_product_measurement"` or `"ppm"`.

alexanderivrii replied:
I thought that we should to use the camel case name because Pauli evolution gates are called `"PauliEvolution"`. But I agree, using snake case naming is more consistent, so it's renamed to `"pauli_product_measurement"` in [388e3d9](https://github.com/Qiskit/qiskit/pull/15126/commits/388e3d98025af0f85edfa4318cd6ce0fd3588ada).

mtreinish commented on crates/circuit/src/circuit_instruction.rs:
```diff
@@ -618,6 +622,51 @@ impl<'a, 'py> FromPyObject<'a, 'py> for OperationFromPython {
+                    "invalid input: {}",
+                    ob.to_owned()
+                )));
+            };
```
I would say instead of params we should leave these as attributes of the Python class and not jam them in params. On unitary gate the params is a historical artifact that would be painful to change, but on a new instruction class we don't have to populate parameters and I think it would be better to use `getattr()` here to access named attributes. This has two advantages, first it doesn't imply you can easily change the pauli and not change the object but more importantly you don't have to explain why changing the parameters directly in place doesn't reflect in all cases.

alexanderivrii replied:
Done in [9beef61](https://github.com/Qiskit/qiskit/pull/15126/commits/9beef612d7c25d8dea24f097ed4478e7aacb93f6). When I implemented this, I treated the implementation for `UnitaryGate` as the source of truth, and did not realize that this is more of an historical artifact. :)

mtreinish commented on crates/circuit/src/operations.rs:
```diff
@@ -2896,3 +2909,84 @@ impl UnitaryGate {
+
+impl Operation for PauliProductMeasurement {
+    fn name(&self) -> &str {
+        "PauliProductMeasurement"
```
As I said before this is a bit odd as a name, normally we don't use camel case or the class name.

alexanderivrii replied:
Done.

mtreinish commented on crates/circuit/src/operations.rs:
```diff
@@ -2896,3 +2909,84 @@ impl UnitaryGate {
+impl PauliProductMeasurement {
+    pub fn create_py_op(&self, py: Python, label: Option<&str>) -> PyResult<Py<PyAny>> {
+        let z = PyList::new(py, &self.z)?;
+        let x = PyList::new(py, &self.x)?;
```
Is this a list or a numpy array in the python api (I'll see when I get to the code). But I remember in the Python -> rust side it was extracting a 1d array and converting that to a vec. Would we avoid some overhead here if we just went to numpy array directly?

alexanderivrii replied:
Yes, it is the numpy array in Python. In [bfbe0f1](https://github.com/Qiskit/qiskit/pull/15126/commits/bfbe0f1fabedb0606bf0c5ef18b7b1a11c2a42f1) I have changed this to 
```rust
let z = self.z.clone().into_bound_py_any(py)?;
let x = self.x.clone().into_bound_py_any(py)?;
```
Is there a way I could avoid cloning the data?

mtreinish replied:
You still need to copy, since we're not giving up ownership of `self.z` or `self.x`. But what you changed it to still results in two copies. `into_bound_py_any()` results in copying into a PyList, then you give that to python which will hand it off to numpy to make the array in the python object. Since this merged already, I pushed up: https://github.com/Qiskit/qiskit/pull/15310 to fix this and reduce it to one copy.

alexanderivrii committed:
Merge branch 'main' into ppm

alexanderivrii committed:
renaming the name to snake case

alexanderivrii committed:
storing data internally rather than on params

alexanderivrii committed:
avoid converting to list

Cryoris commented on crates/circuit/src/circuit_instruction.rs:
```diff
@@ -618,6 +622,32 @@ impl<'a, 'py> FromPyObject<'a, 'py> for OperationFromPython {
+            let pauli_product_measurement = Box::new(PauliProductMeasurement {
+                z: z.to_owned(),
+                x: x.to_owned(),
+                neg: phase == 2,
```
It might be worth adding a clarifying comment here that explains that `phase` can only be 0 or 2 from input validation

ShellyGarion replied:
done in [38766b1](https://github.com/Qiskit/qiskit/pull/15126/commits/38766b10660866e8456c217f2206fc5dc8c2beb2)

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,157 @@
+        circuit = QuantumCircuit._from_circuit_data(
+            synth_pauli_product_measurement(self),
+            legacy_qubits=True,
+            name="ppm_circuit",
```
from our offline discussions, maybe this? 🙂 
```suggestion
            name="def_ppm",
```

Cryoris commented on crates/synthesis/src/pauli_product_measurement.rs:
```diff
@@ -0,0 +1,134 @@
+        )?;
+    }
+
+    // Basis change layer
```
```suggestion
    // Flip the sign of the measurement outcome, if specified
```

Cryoris commented on crates/synthesis/src/pauli_product_measurement.rs:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    if ppm.neg {
```
I would've expected the negation to sandwich the single Z measurement, i.e.
```
basis change
CX chain
if neg { X }
Z measure
if neg { X }
inv(CX chain)
inv(basis change)
```
Even if it would happen to be equivalent since the `first_qubit` happens to never act as control in the CX chain, I think it would be clearer to move it inside

ShellyGarion replied:
done in [aa1982e](https://github.com/Qiskit/qiskit/pull/15126/commits/aa1982e265e3d705db537652c9ac0fc33f3d3b29)

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,157 @@
+        if not isinstance(other, PauliProductMeasurement):
+            return False
+
+        if self.label != other.label:
```
This currently does redundant checks in the default path since the default label specifies the Paulis already. One potential solution would be to add a `label` property that returns `_default_label` if the user-specified label is `None`.

ShellyGarion replied:
don't we do it in lines 91-92?
```
       if label is None:
            label = _get_default_label(pauli)
```

alexanderivrii replied:
Here I followed the code for `PauliEvolutionGate` which also sets `self.label` in the constructor (to the default label, if `None`). I don't think it poses much of an overhead since this code is not called from rust.

Cryoris replied:
Shelly: I mean that in the equality check you have
```python
if self.label != other.label:  # per default the labels equal the Pauli string

if self.x == other.x and self.z == other.z and ... # again comparing the Pauli strings
```

Sasha: `PauliEvolutionGate` doesn't use the label for equality comparisons though, or does it?

alexanderivrii replied:
`UnitaryGate.__eq__` also compares labels in the Python code, so I thought it's the correct thing to do.

ShellyGarion replied:
removed check in [3fef300](https://github.com/Qiskit/qiskit/pull/15126/commits/3fef3007c23df60c4ef6689ea98590c318daff53)

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,157 @@
+    return "PPM(" + pauli.to_label() + ")"
+
+
+def _is_empty_pauli(pauli: Pauli):
```
This is a bit subjective but this method seems a bit redundant given that we could just replace calling it with `len(pauli) == 0` -- but if you think this simplifies the code structure I'm also fine keeping it

ShellyGarion replied:
done in [84d8629](https://github.com/Qiskit/qiskit/pull/15126/commits/84d862917a845d4f1f642fdb02a511f51751c425)

Cryoris commented on qiskit/circuit/library/pauli_product_measurement.py:
```diff
@@ -0,0 +1,157 @@
+
+
+def _get_default_label(pauli: Pauli):
+    """Creates the detault label for PauliProductMeasurement instruction,
```
```suggestion
    """Creates the default label for PauliProductMeasurement instruction,
```

Cryoris commented on releasenotes/notes/add-pauli-product-measurement-instruction-49abd4141328b46d.yaml:
```diff
@@ -0,0 +1,29 @@
+    Added a new instruction class, :class:`.PauliProductMeasurement`, which represents
+    a fundamental operation in fault-tolerant quantum computing. 
+    Mathematically, it performs a joint projective measurement on
+    multiple qubits, where the measured observable is a tensor product of Pauli operators.
```
We can simplify this a bit (also in the class docstring):
```suggestion
    Added a new instruction class, :class:`.PauliProductMeasurement`, which represents
    a joint projective measurement on multiple qubits, where the measured observable is 
    a tensor product of Pauli operators.
```

Cryoris commented on releasenotes/notes/add-pauli-product-measurement-instruction-49abd4141328b46d.yaml:
```diff
@@ -0,0 +1,29 @@
+    "A Game of Surface Codes: Large-Scale Quantum Computing with Lattice Surgery" by
+    Daniel Litinski (https://arxiv.org/abs/1808.02892).
+
+    A :class:`.PauliProducMeasurement` can be instrantiated from a 
```
```suggestion
    A :class:`.PauliProductMeasurement` can be instantiated from a 
```

Cryoris reviewed (commented):
Some minor comments, but overall LGTM. We can improve the efficiency of the internal representation later on, or if we notice it becomes a bottleneck.

ShellyGarion committed:
Update name to def_ppm

Co-authored-by: Julien Gacon <gaconju@gmail.com>

ShellyGarion committed:
Update comment on layer

Co-authored-by: Julien Gacon <gaconju@gmail.com>

ShellyGarion committed:
Fix misprint

Co-authored-by: Julien Gacon <gaconju@gmail.com>

ShellyGarion committed:
Simplify release notes

Co-authored-by: Julien Gacon <gaconju@gmail.com>

ShellyGarion committed:
Fix misprint

Co-authored-by: Julien Gacon <gaconju@gmail.com>

ShellyGarion committed:
add comment on phase

ShellyGarion committed:
move neg

ShellyGarion committed:
remove _is_empty_pauli func

ShellyGarion committed:
remove unnecessary label checks

Cryoris reviewed (dismissed):
LGTM, thanks for all the work!

There are some follow ups for updating measurement-related transpiler passes, but this looks in a good shape already to unblock the FTQC demo and the Litinski transformation.

alexanderivrii committed:
Merge branch 'main' into ppm

alexanderivrii committed:
following an update, adding missing PPM variant to cext/dag.rs
